File: tests/oracle/action_lanedivide_1204.cc
/* ACTION-LANEDIVIDE-0001: locked Ghidra 12.0.4 oracle fixture.
 *
 * Exercises the real ActionLaneDivide::apply (coreaction.cc:585-622) —
 * collectLaneSizes candidate filtering, processVarnode success (mode 0
 * collected lanes) and failure (mode 2 default pointer-size lane with a
 * non-splittable INT_MULT definition), the post-apply IR and change
 * counter, the rule_onceperfunc second-pass skip, and the
 * clearLanedAccessMap timing.
 */
#include "bfd_arch.hh"
#include "coreaction.hh"
#include "libdecomp.hh"

#include <iostream>
#include <iterator>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::map;
using std::ostringstream;
using std::string;
using std::vector;

/* Exposes the protected Action bookkeeping (count/status) for projection. */
class ProbeLaneDivide : public ActionLaneDivide {
public:
  ProbeLaneDivide(const string &g) : ActionLaneDivide(g) {}
  int4 exposedCount() const { return count; }
  uint4 exposedStatus() const { return status; }
  virtual Action *clone(const ActionGroupList &grouplist) const {
    if (!grouplist.contains(getGroup())) return (Action *)0;
    return new ProbeLaneDivide(getGroup());
  }
};

struct GraphProjection {
  vector<PcodeOp *> ops;
  map<PcodeOp *,int4> opIndex;
  vector<Varnode *> vars;
  map<Varnode *,int4> varIndex;

  explicit GraphProjection(BlockBasic *block)
  {
    for(list<PcodeOp *>::const_iterator iter=block->beginOp();iter!=block->endOp();++iter) {
      opIndex[*iter] = ops.size();
      ops.push_back(*iter);
    }
    for(vector<PcodeOp *>::const_iterator iter=ops.begin();iter!=ops.end();++iter) {
      PcodeOp *op = *iter;
      touch(op->getOut());
      for(int4 slot=0;slot<op->numInput();++slot)
        touch(op->getIn(slot));
    }
  }

  void touch(Varnode *vn)
  {
    if (vn == (Varnode *)0 || varIndex.find(vn) != varIndex.end()) return;
    varIndex[vn] = vars.size();
    vars.push_back(vn);
  }

  string varName(Varnode *vn) const
  {
    if (vn == (Varnode *)0) return "_";
    map<Varnode *,int4>::const_iterator iter = varIndex.find(vn);
    if (iter == varIndex.end()) return "x";
    ostringstream out;
    out << 'v' << (*iter).second;
    return out.str();
  }

  string opName(PcodeOp *op) const
  {
    if (op == (PcodeOp *)0) return "_";
    map<PcodeOp *,int4>::const_iterator iter = opIndex.find(op);
    if (iter == opIndex.end()) return "x";
    ostringstream out;
    out << 'o' << (*iter).second;
    return out.str();
  }

  string render(const Funcdata *fd,BlockBasic *block) const
  {
    ostringstream out;
    out << "ops[";
    for(int4 i=0;i<ops.size();++i) {
      if (i != 0) out << ';';
      PcodeOp *op = ops[i];
      out << 'o' << i << ':' << (int4)op->code()
          << "@" << op->getAddr().getOffset()
          << "/t" << op->getSeqNum().getTime()
          << "/r" << op->getSeqNum().getOrder()
          << "/d" << (op->isDead() ? 1 : 0)
          << "/p" << (op->getParent() == block ? block->getIndex() : -1)
          << "/o" << varName(op->getOut()) << "/i";
      for(int4 slot=0;slot<op->numInput();++slot) {
        if (slot != 0) out << ',';
        out << varName(op->getIn(slot));
      }
    }
    out << "]vars[";
    for(int4 i=0;i<vars.size();++i) {
      if (i != 0) out << ';';
      Varnode *vn = vars[i];
      out << 'v' << i << ":c" << vn->getCreateIndex()
          << "/s" << vn->getSize()
          << "/sp" << vn->getSpace()->getIndex()
          << "/k" << (vn->isConstant() ? 1 : 0);
      if (vn->isConstant()) out << ':' << vn->getOffset();
      out << "/f" << (vn->isFree() ? 1 : 0)
          << "/n" << (vn->isInput() ? 1 : 0)
          << "/w" << (vn->isWritten() ? 1 : 0)
          << "/d" << opName(vn->getDef()) << "/u";
      bool first = true;
      for(list<PcodeOp *>::const_iterator iter=vn->beginDescend();iter!=vn->endDescend();++iter) {
        if (!first) out << ',';
        first = false;
        out << opName(*iter);
      }
    }
    out << "]count=" << ops.size()
        << ',' << vars.size()
        << ',' << std::distance(fd->beginOpAlive(),fd->endOpAlive())
        << ',' << std::distance(fd->beginOpDead(),fd->endOpDead())
        << ',' << std::distance(fd->beginOpAll(),fd->endOpAll())
        << ',' << fd->numVarnodes();
    return out.str();
  }
};

string snapshot(const Funcdata *fd,BlockBasic *block)
{
  GraphProjection projection(block);
  return projection.render(fd,block);
}

BlockBasic *newBlock(Funcdata *fd)
{
  BlockGraph &graph = const_cast<BlockGraph &>(fd->getBasicBlocks());
  return graph.newBlockBasic(fd);
}

PcodeOp *newOutputOp(Funcdata *fd,BlockBasic *block,OpCode opcode,uintb pc,int4 inputs,int4 outputSize)
{
  Address address(fd->getArch()->getDefaultCodeSpace(),pc);
  PcodeOp *op = fd->newOp(inputs,address);
  fd->opSetOpcode(op,opcode);
  fd->newUniqueOut(outputSize,op);
  fd->opInsertEnd(op,block);
  return op;
}

string laneMap(const Funcdata *fd,const Architecture &architecture)
{
  ostringstream out;
  bool first = true;
  for(map<VarnodeData,const LanedRegister *>::const_iterator iter=fd->beginLaneAccess();
      iter!=fd->endLaneAccess();++iter) {
    if (!first) out << ';';
    first = false;
    const VarnodeData &storage = (*iter).first;
    const LanedRegister *record = (*iter).second;
    int4 identity = -1;
    for(int4 i=0;i<architecture.lanerecords.size();++i) {
      if (record == &architecture.lanerecords[i]) {
        identity = i;
        break;
      }
    }
    out << storage.space->getIndex() << ':' << storage.offset << ':' << storage.size
        << "=r" << identity << ':' << record->getWholeSize()
        << ':' << record->getSizeBitMask();
  }
  return out.str();
}

/* Success path: 4-byte PIECE root with two 2-byte SUBPIECE descendants.
 * collectLaneSizes accepts lane size 2 from the descendants and from the
 * PIECE definition; processVarnode mode 0 splits; the second perform must
 * skip apply entirely (rule_onceperfunc), leaving a re-queued laned-map
 * entry intact. */
void runPiece(BfdArchitecture &architecture)
{
  Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("main_free");
  if (fd == (Funcdata *)0) throw std::runtime_error("main_free not found");
  BlockBasic *block = newBlock(fd);
  PcodeOp *piece = newOutputOp(fd,block,CPUI_PIECE,0x5000,2,4);
  Varnode *root = piece->getOut();
  fd->opSetInput(piece,fd->newConstant(2,0x1122),0);
  fd->opSetInput(piece,fd->newConstant(2,0x3344),1);
  PcodeOp *low = newOutputOp(fd,block,CPUI_SUBPIECE,0x5001,2,2);
  fd->opSetInput(low,root,0);
  fd->opSetInput(low,fd->newConstant(4,0),1);
  PcodeOp *high = newOutputOp(fd,block,CPUI_SUBPIECE,0x5002,2,2);
  fd->opSetInput(high,root,0);
  fd->opSetInput(high,fd->newConstant(4,2),1);
  string mapBefore = laneMap(fd,architecture);
  string irBefore = snapshot(fd,block);
  ProbeLaneDivide action("probe");
  int4 ret1 = action.perform(*fd);
  string mapAfter = laneMap(fd,architecture);
  string irAfter = snapshot(fd,block);
  AddrSpace *registerSpace = architecture.getSpaceByName("register");
  fd->checkForLanedRegister(16,Address(registerSpace,0x77));
  string mapRequeued = laneMap(fd,architecture);
  int4 ret2 = action.perform(*fd);
  string mapAfterSecond = laneMap(fd,architecture);
  string irAfterSecond = snapshot(fd,block);
  std::cout << "piece|ret1=" << ret1
            << "|count=" << action.exposedCount()
            << "|status=" << action.exposedStatus()
            << "|map=" << mapBefore << '>' << mapAfter
            << "|requeued=" << mapRequeued
            << "|ret2=" << ret2
            << "|map2=" << mapAfterSecond
            << "|status2=" << action.exposedStatus()
            << "|count2=" << action.exposedCount()
            << "|irStable=" << (irAfter == irAfterSecond ? 1 : 0)
            << "|before=" << irBefore
            << "|after=" << irAfter << '\n';
}

/* Failure path: 16-byte INT_MULT root (INT_MULT is not lane-splittable
 * backward), one 2-byte SUBPIECE descendant. collectLaneSizes rejects
 * lane size 2 against the whole-size-16 record (bit 8 only); modes 0/1
 * collect nothing; mode 2 falls back to the default pointer-size lane 8
 * and the backward trace rejects INT_MULT. Zero mutation; the laned map
 * is still cleared at the end of apply. */
void runFailure(BfdArchitecture &architecture)
{
  Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("main_init");
  if (fd == (Funcdata *)0) throw std::runtime_error("main_init not found");
  BlockBasic *block = newBlock(fd);
  PcodeOp *multiply = newOutputOp(fd,block,CPUI_INT_MULT,0x7000,2,16);
  Varnode *root = multiply->getOut();
  fd->opSetInput(multiply,fd->newConstant(16,0x1122334455667788),0);
  fd->opSetInput(multiply,fd->newConstant(16,0x99aabbccddeeff00),1);
  PcodeOp *low = newOutputOp(fd,block,CPUI_SUBPIECE,0x7001,2,2);
  fd->opSetInput(low,root,0);
  fd->opSetInput(low,fd->newConstant(4,0),1);
  string mapBefore = laneMap(fd,architecture);
  string irBefore = snapshot(fd,block);
  ProbeLaneDivide action("probe");
  int4 ret1 = action.perform(*fd);
  string mapAfter = laneMap(fd,architecture);
  string irAfter = snapshot(fd,block);
  std::cout << "failure|ret1=" << ret1
            << "|count=" << action.exposedCount()
            << "|status=" << action.exposedStatus()
            << "|map=" << mapBefore << '>' << mapAfter
            << "|irSame=" << (irBefore == irAfter ? 1 : 0)
            << "|before=" << irBefore
            << "|after=" << irAfter << '\n';
}

void run(const string &specDirectory,const string &binary)
{
  vector<string> specPaths(1,specDirectory);
  startDecompilerLibrary(specPaths);
  {
    ostringstream messages;
    BfdArchitecture architecture(binary,"default",&messages);
    DocumentStorage store;
    architecture.init(store);
    architecture.lanerecords.clear();
    architecture.lanerecords.push_back(LanedRegister(4,1u << 2));
    architecture.lanerecords.push_back(LanedRegister(16,1u << 8));
    architecture.readLoaderSymbols("::");
    // Project the normalized mode-2 default lane size (what
    // processVarnode actually consumes), not the raw pointer size: the
    // C++ oracle reads a spec-configured TypeFactory while the Rust
    // comparand drives a bare Architecture whose TypeFactory is
    // unconfigured; coreaction.cc:566-569 normalizes both to the same
    // default lane.
    int4 defaultSize = architecture.types->getSizeOfPointer();
    if (defaultSize != 4) defaultSize = 8;
    std::cout << "arch|default=" << defaultSize
              << "|min=" << architecture.getMinimumLanedRegisterSize()
              << "|sizes=" << architecture.lanerecords[0].getWholeSize()
              << ',' << architecture.lanerecords[1].getWholeSize() << '\n';
    runPiece(architecture);
    runFailure(architecture);
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(int argc,char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: action_lanedivide_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    run(argv[1],argv[2]);
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/action_lanedivide_1204.rs
//! ACTION-LANEDIVIDE-0001 Rugra comparand.
//!
//! Drives the real `ActionLaneDivide` (rugra::coreaction) through the
//! public `Action::perform` state machine, mirroring the locked-oracle
//! C++ fixture observation for observation.

use rugra::action::{Action, ActionState, action_flags};
use rugra::address::Address;
use rugra::arch::Architecture;
use rugra::block::FlowBlock;
use rugra::coreaction::ActionLaneDivide;
use rugra::funcdata::Funcdata;
use rugra::op::PcodeOpRef;
use rugra::opcodes::OpCode;
use rugra::space::AddressSpace;
use rugra::transform::LanedRegister;
use rugra::varnode::Varnode;
use std::collections::HashMap;
use std::fmt::Write;
use std::sync::{Arc, RwLock};

type Block = Arc<RwLock<dyn FlowBlock + Send + Sync>>;

struct GraphProjection {
    ops: Vec<PcodeOpRef>,
    op_index: HashMap<usize, usize>,
    vars: Vec<Arc<RwLock<Varnode>>>,
    var_index: HashMap<usize, usize>,
}

impl GraphProjection {
    fn new(block: &Block) -> Self {
        let ops = block.read().unwrap().get_ops();
        let op_index = ops
            .iter()
            .enumerate()
            .map(|(index, op)| (Arc::as_ptr(&op.0) as usize, index))
            .collect();
        let mut result = Self {
            ops,
            op_index,
            vars: Vec::new(),
            var_index: HashMap::new(),
        };
        for op_index in 0..result.ops.len() {
            let (output, inputs) = {
                let op = result.ops[op_index].0.read().unwrap();
                (op.get_out().cloned(), op.inrefs.clone())
            };
            result.touch(output);
            for input in inputs {
                result.touch(Some(input));
            }
        }
        result
    }

    fn touch(&mut self, varnode: Option<Arc<RwLock<Varnode>>>) {
        let Some(varnode) = varnode else { return };
        let pointer = Arc::as_ptr(&varnode) as usize;
        if self.var_index.contains_key(&pointer) {
            return;
        }
        self.var_index.insert(pointer, self.vars.len());
        self.vars.push(varnode);
    }

    fn var_name(&self, varnode: Option<&Arc<RwLock<Varnode>>>) -> String {
        let Some(varnode) = varnode else {
            return "_".to_string();
        };
        self.var_index
            .get(&(Arc::as_ptr(varnode) as usize))
            .map_or_else(|| "x".to_string(), |index| format!("v{index}"))
    }

    fn op_name(&self, op: Option<&Arc<RwLock<rugra::op::PcodeOp>>>) -> String {
        let Some(op) = op else {
            return "_".to_string();
        };
        self.op_index
            .get(&(Arc::as_ptr(op) as usize))
            .map_or_else(|| "x".to_string(), |index| format!("o{index}"))
    }

    fn render(&self, fd: &Funcdata, block: &Block) -> String {
        let mut output = String::new();
        output.push_str("ops[");
        for (index, op) in self.ops.iter().enumerate() {
            if index != 0 {
                output.push(';');
            }
            let op = op.0.read().unwrap();
            let parent = op
                .parent
                .as_ref()
                .and_then(std::sync::Weak::upgrade)
                .filter(|parent| Arc::ptr_eq(parent, block))
                .map_or(-1, |parent| parent.read().unwrap().get_index());
            write!(
                output,
                "o{index}:{}@{}/t{}/r{}/d{}/p{parent}/o{}/i",
                op.opcode as i32,
                op.get_addr().as_u64(),
                op.start.get_time(),
                op.start.get_order(),
                u8::from(op.is_dead()),
                self.var_name(op.get_out()),
            )
            .unwrap();
            for (slot, input) in op.inrefs.iter().enumerate() {
                if slot != 0 {
                    output.push(',');
                }
                output.push_str(&self.var_name(Some(input)));
            }
        }
        output.push_str("]vars[");
        for (index, varnode) in self.vars.iter().enumerate() {
            if index != 0 {
                output.push(';');
            }
            let varnode = varnode.read().unwrap();
            write!(
                output,
                "v{index}:c{}/s{}/sp{}/k{}",
                varnode.create_index,
                varnode.get_size(),
                varnode.get_space().space_id(),
                u8::from(varnode.is_constant()),
            )
            .unwrap();
            if varnode.is_constant() {
                write!(output, ":{}", varnode.get_offset()).unwrap();
            }
            write!(
                output,
                "/f{}/n{}/w{}/d{}/u",
                u8::from(varnode.is_free()),
                u8::from(varnode.is_input()),
                u8::from(varnode.is_written()),
                self.op_name(varnode.get_def().as_ref()),
            )
            .unwrap();
            for (use_index, descendant) in varnode.descend_iter().enumerate() {
                if use_index != 0 {
                    output.push(',');
                }
                output.push_str(&self.op_name(Some(&descendant)));
            }
        }
        write!(
            output,
            "]count={},{},{},{},{},{}",
            self.ops.len(),
            self.vars.len(),
            fd.obank.alivelist.len(),
            fd.obank.deadlist.len(),
            fd.obank.optree.len(),
            fd.vbank.num_varnodes(),
        )
        .unwrap();
        output
    }
}

fn snapshot(fd: &Funcdata, block: &Block) -> String {
    GraphProjection::new(block).render(fd, block)
}

fn new_output_op(
    fd: &mut Funcdata,
    block: &Block,
    opcode: OpCode,
    pc: u64,
    inputs: usize,
    output_size: usize,
) -> (PcodeOpRef, Arc<RwLock<Varnode>>) {
    let op = fd.new_op(inputs, Address::new(pc));
    fd.op_set_opcode(&op, opcode);
    let output = fd.new_unique_out(output_size, &op);
    fd.op_insert_end(&op, block);
    (op, output)
}

fn lane_map(fd: &Funcdata, architecture: &Architecture) -> String {
    fd.lane_accesses()
        .map(|(storage, record)| {
            let identity = architecture
                .lane_records
                .iter()
                .position(|candidate| Arc::ptr_eq(candidate, record))
                .map_or(-1, |index| index as i32);
            format!(
                "{}:{}:{}=r{}:{}:{}",
                storage.space.space_id(),
                storage.offset,
                storage.size,
                identity,
                record.get_whole_size(),
                record.get_size_bit_mask(),
            )
        })
        .collect::<Vec<_>>()
        .join(";")
}

fn new_function(name: &str, address: u64, architecture: &Arc<Architecture>) -> (Funcdata, Block) {
    let mut fd = Funcdata::new(name, Address::new(address), 0);
    fd.set_arch(architecture.clone());
    let block = fd.create_new_block();
    (fd, block)
}

fn run_piece(architecture: &Arc<Architecture>) {
    let (mut fd, block) = new_function("main_free", 0x4970, architecture);
    let (piece, root) = new_output_op(&mut fd, &block, OpCode::CPUI_PIECE, 0x5000, 2, 4);
    let high_input = fd.new_constant(2, 0x1122);
    let low_input = fd.new_constant(2, 0x3344);
    fd.op_set_input(&piece, high_input, 0);
    fd.op_set_input(&piece, low_input, 1);
    let (low, _) = new_output_op(&mut fd, &block, OpCode::CPUI_SUBPIECE, 0x5001, 2, 2);
    fd.op_set_input(&low, root.clone(), 0);
    let zero = fd.new_constant(4, 0);
    fd.op_set_input(&low, zero, 1);
    let (high, _) = new_output_op(&mut fd, &block, OpCode::CPUI_SUBPIECE, 0x5002, 2, 2);
    fd.op_set_input(&high, root.clone(), 0);
    let two = fd.new_constant(4, 2);
    fd.op_set_input(&high, two, 1);
    let map_before = lane_map(&fd, architecture);
    let ir_before = snapshot(&fd, &block);
    let mut action = ActionLaneDivide::new();
    let mut state = ActionState::new(action_flags::RULE_ONCEPERFUNC);
    let ret1 = action
        .perform(&mut fd, &mut state)
        .expect("first perform");
    let map_after = lane_map(&fd, architecture);
    let ir_after = snapshot(&fd, &block);
    fd.check_for_laned_register(16, AddressSpace::Register, Address::new(0x77));
    let map_requeued = lane_map(&fd, architecture);
    let ret2 = action
        .perform(&mut fd, &mut state)
        .expect("second perform");
    let map_after_second = lane_map(&fd, architecture);
    let ir_after_second = snapshot(&fd, &block);
    println!(
        "piece|ret1={ret1}|count={}|status={}|map={map_before}>{map_after}|requeued={map_requeued}|ret2={ret2}|map2={map_after_second}|status2={}|count2={}|irStable={}|before={ir_before}|after={ir_after}",
        state.count,
        state.status,
        state.status,
        state.count,
        u8::from(ir_after == ir_after_second),
    );
}

fn run_failure(architecture: &Arc<Architecture>) {
    let (mut fd, block) = new_function("main_init", 0x4960, architecture);
    let (multiply, root) = new_output_op(&mut fd, &block, OpCode::CPUI_INT_MULT, 0x7000, 2, 16);
    let left = fd.new_constant(16, 0x1122334455667788);
    let right = fd.new_constant(16, 0x99aabbccddeeff00);
    fd.op_set_input(&multiply, left, 0);
    fd.op_set_input(&multiply, right, 1);
    let (low, _) = new_output_op(&mut fd, &block, OpCode::CPUI_SUBPIECE, 0x7001, 2, 2);
    fd.op_set_input(&low, root.clone(), 0);
    let zero = fd.new_constant(4, 0);
    fd.op_set_input(&low, zero, 1);
    let map_before = lane_map(&fd, architecture);
    let ir_before = snapshot(&fd, &block);
    let mut action = ActionLaneDivide::new();
    let mut state = ActionState::new(action_flags::RULE_ONCEPERFUNC);
    let ret1 = action
        .perform(&mut fd, &mut state)
        .expect("failure perform");
    let map_after = lane_map(&fd, architecture);
    let ir_after = snapshot(&fd, &block);
    println!(
        "failure|ret1={ret1}|count={}|status={}|map={map_before}>{map_after}|irSame={}|before={ir_before}|after={ir_after}",
        state.count,
        state.status,
        u8::from(ir_before == ir_after),
    );
}

fn main() {
    let mut architecture = Architecture::new();
    architecture.set_lane_records(vec![
        LanedRegister::with_sizes(4, 1 << 2),
        LanedRegister::with_sizes(16, 1 << 8),
    ]);
    let architecture = Arc::new(architecture);
    // Project the normalized mode-2 default lane size (what
    // process_varnode actually consumes, mirroring coreaction.cc:566-569),
    // not the raw pointer size: the C++ oracle reads a spec-configured
    // TypeFactory while this comparand drives a bare Architecture whose
    // TypeFactory is unconfigured; the !=4 normalization maps both to the
    // same default lane.
    let mut default_size = architecture
        .types
        .as_ref()
        .map(|types| types.read().unwrap().get_size_of_pointer())
        .unwrap_or(0);
    if default_size != 4 {
        default_size = 8;
    }
    println!(
        "arch|default={default_size}|min={}|sizes={},{}",
        architecture.get_minimum_laned_register_size(),
        architecture.lane_records[0].get_whole_size(),
        architecture.lane_records[1].get_whole_size(),
    );
    run_piece(&architecture);
    run_failure(&architecture);
}
